### PR TITLE
Sample terminology

### LINKED ISSUE #225: "tracked leaves" should be "tracked samples"
As mentioned in #223, to change the name, keeping these as deprecated aliases:

- `tracked_leaves`
- `leaf_counts`
- `leaf_lists`

To *not* change:

- `is_leaf` and `is_internal`

but perhaps we should add a `is_sample` function.

### PR DESCRIPTION
**Pull request by jeromekelleher**
Closes #225. Also adds much better support for internal samples in the test cases.

TODO: Some work to do with simplify on internal sample examples.

### CODE REVIEW

**Comment by jeromekelleher** [comment]
Would you mind having a quick look over this @petrelharp or @ashander?  It's a bit change, and I've basically just done s/leaf/sample everywhere, so it would be great to get another pair of eyes on the diff.

File: lib/tests.c [hunk 1]
```diff
@@ -2646,7 +2646,7 @@ test_simplest_unary_records(void)
 }
 
 static void
-test_simplest_non_sample_leaf_records(void)
+test_simplest_non_sample_sample_records(void)
```
Comment (petrelharp) [inline lib/tests.c]: I think this should still be as before, "test_simplest_non_sample_leaf_records"

File: lib/tests.c [hunk 2]
```diff
@@ -7265,7 +7268,7 @@ main(int argc, char **argv)
         {"test_simplest_records", test_simplest_records},
         {"test_simplest_nonbinary_records", test_simplest_nonbinary_records},
         {"test_simplest_unary_records", test_simplest_unary_records},
-        {"test_simplest_non_sample_leaf_records", test_simplest_non_sample_leaf_records},
+        {"test_simplest_non_sample_sample_records", test_simplest_non_sample_sample_records},
```
Comment (petrelharp) [inline lib/tests.c]: also, here.

File: msprime/trees.py [hunk 3]
```diff
@@ -684,7 +684,7 @@ def time(self, u):
     def get_time(self, u):
         """
         Returns the time of the specified node in generations. Returns 0 if u
-        is a leaf or is not a node in the current tree.
+        is a sample or is not a node in the current tree.
```
Comment (petrelharp) [inline msprime/trees.py]: I think that this should read "Returns 0 if u is not a node in the current tree."

File: msprime/trees.py [hunk 4]
```diff
@@ -669,7 +669,7 @@ def get_children(self, u):
         """
         Returns the children of the specified node as a tuple :math:`(v, w)`.
         For internal nodes, this tuple is always in sorted order such that
-        :math:`v < w`. If u is a leaf or is not a node in the current tree,
+        :math:`v < w`. If u is a sample or is not a node in the current tree,
```
Comment (petrelharp) [inline msprime/trees.py]: I think this should read "if u is a leaf".

File: msprime/trees.py [hunk 5]
```diff
@@ -697,7 +697,7 @@ def population(self, u):
 
     def get_population(self, u):
         """
-        Returns the population associated with the specified node. For leaf
+        Returns the population associated with the specified node. For sample
```
Comment (petrelharp) [inline msprime/trees.py]: I am not sure this distinction needs to be made?  If not, could delete this sentence.

File: msprime/trees.py [hunk 6]
```diff
@@ -2246,7 +2311,7 @@ def diffs(self):
         tuples, and these describe the records that must be applied to create
         the tree covering the current interval. These records are returned in
         time-increasing order, such that the records affecting the lowest parts
-        of the tree (i.e., closest to the leaves) are returned first.
+        of the tree (i.e., closest to the samples) are returned first.
```
Comment (petrelharp) [inline msprime/trees.py]: how about instead of "closest to the leaves" say "closest to the present day"

Comment (petrelharp) [reply msprime/trees.py]: or if not, "closest to the leaves"

File: _msprimemodule.c [hunk 7]
```diff
@@ -122,9 +122,9 @@ typedef struct {
 typedef struct {
     PyObject_HEAD
     SparseTree *sparse_tree;
-    leaf_list_node_t *head;
-    leaf_list_node_t *tail;
-    leaf_list_node_t *next;
+    node_list_t *head;
+    node_list_t *tail;
+    node_list_t *next;
 } LeafListIterator;
```
Comment (petrelharp) [inline _msprimemodule.c]: By my reading this should be `SampleListIterator`, but this is never externally visible, so not a big deal.

File: tests/test_topology.py [hunk 8]
```diff
@@ -1094,7 +1094,7 @@ def test_simplest_non_degenerate_case(self):
         self.assertEqual(t.time_dict, {0: 0, 1: 0, 2: 0, 3: 0, 4: 1, 5: 2})
 
     def test_two_reducable_trees(self):
-        # We have n = 4 and two trees, with some unary nodes and non-sample leaves
+        # We have n = 4 and two trees, with some unary nodes and non-sample samples
```
Comment (petrelharp) [inline tests/test_topology.py]: should be "non-sample leaves"

File: tests/test_topology.py [hunk 9]
```diff
@@ -1105,7 +1105,7 @@ def test_two_reducable_trees(self):
         5       0           1
         6       0           2
         7       0           3
-        8       0           0   # Non sample leaf
+        8       0           0   # Non sample sample
```
Comment (petrelharp) [inline tests/test_topology.py]: Non sample leaf

File: tests/test_topology.py [hunk 10]
```diff
@@ -1177,7 +1177,7 @@ def test_one_reducable_tree(self):
...
```
Comment (petrelharp) [inline tests/test_topology.py]: Again

File: tests/test_topology.py [hunk 11]
```diff
@@ -1357,7 +1357,7 @@ def test_partial_non_sample_external_nodes(self):
         0       1           0
         1       1           0
         2       1           0
-        3       0           0.2  # Non sample leaf
+        3       0           0.2  # Non sample sample
```
Comment (petrelharp) [inline tests/test_topology.py]: Again

File: tests/test_topology.py [hunk 12]
```diff
@@ -1415,7 +1415,7 @@ def test_partial_non_sample_external_nodes_2(self):
         4       0           0.5
         5       0           0.7
         6       0           1.0
-        7       0           0    # Non sample leaf
+        7       0           0    # Non sample sample
```
Comment (petrelharp) [inline tests/test_topology.py]: Again

File: tests/test_topology.py [hunk 13]
```diff
@@ -1468,7 +1468,7 @@ def test_single_offspring_records(self):
         0   1           0
         1   1           0
         2   1           0
-        3   0           0       # Non sample leaf
+        3   0           0       # Non sample sample
```
Comment (petrelharp) [inline tests/test_topology.py]: Again

**Comment by petrelharp** [comment]
Whew, looked through all that.  Now I need to go take a walk under the sampley branches of the trees outside...

File: msprime/trees.py (hunk 3 above)
Comment (jeromekelleher) [reply msprime/trees.py]: Well spotted. Opening an issue on this.

**Comment by jeromekelleher** [comment]
Fixed all those. This was a huge help, thanks @petrelharp! I'm glad we've fixed this terminological mess, it would just have been horribly confusing to try and carry it forward.

**Pull request merged**